pkg/drivers/sqlite/sqlite.go

```diff
@@ -81,27 +63,186 @@ func NewVariant(ctx context.Context, driverName, dataSourceName string) (server.
+// changes are rolled back if an error occurs.
+func setup(dialect *generic.Generic) error {
+	ctx := context.Background()
+	txn, err := dialect.DB.BeginTx(ctx, nil)
+	if err != nil {
+		return err
+	}
+	defer txn.Rollback()
+
+	if err := migration(ctx, txn, dialect); err != nil {
+		return errors.Wrap(err, "migration failed")
+	}
+
+	return txn.Commit()
+}
+
+func createKineTable(txn *sql.Tx) error {
+	createTableSQL := `CREATE TABLE IF NOT EXISTS kine
+			(
+				id INTEGER PRIMARY KEY AUTOINCREMENT,
+				name TEXT NOT NULL,
+				created INTEGER,
+				deleted INTEGER,
+				create_revision INTEGER NOT NULL,
+				prev_revision INTEGER,
+				lease INTEGER,
+				value BLOB,
+				old_value BLOB
+			)`
+
+	_, err := txn.Exec(createTableSQL)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// removeTable drops the table with the given name, in the given transaction.
+func removeTable(txn *sql.Tx, tableName string) error {
+	dropTableSQL := fmt.Sprintf(`DROP TABLE IF EXISTS %s`, tableName)
+	_, err := txn.Exec(dropTableSQL)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// alterIndices drops the given old database indices and creates
+// the given new ones.
+func alterIndices(txn *sql.Tx, d *generic.Generic) error {
+	if d.LockWrites {
+		d.Lock()
+		defer d.Unlock()
+	}
+
+	dropIndices := []string{
+		`DROP INDEX IF EXISTS kine_name_index`,
+		`DROP INDEX IF EXISTS kine_name_prev_revision_uindex`,
+	}
+
+	// Drop the old indices
+	for _, stmt := range dropIndices {
+		_, err := txn.Exec(stmt)
+		if err != nil {
+			return err
+		}
+	}
+
+	createIndices := []string{
+		`CREATE INDEX IF NOT EXISTS kine_name_index ON kine (name, id)`,
+		`CREATE UNIQUE INDEX IF NOT EXISTS kine_name_prev_revision_uindex ON kine (prev_revision, name)`,
+	}
+
+	// Create the new indices
+	for _, stmt := range createIndices {
+		_, err := txn.Exec(stmt)
 		if err != nil {
 			return err
 		}
 	}
 
 	return nil
 }
+
+// countTable counts the number of tables with the given name, in the given transaction.
+func hasTable(ctx context.Context, txn *sql.Tx, tableName string) (bool, error) {
+	tableListSQL := fmt.Sprintf(`SELECT COUNT(*) FROM sqlite_master WHERE type = 'table' AND name = '%s'`, tableName)
+	row := txn.QueryRowContext(ctx, tableListSQL)
+	var tableCount int
+	if err := row.Scan(&tableCount); err != nil {
+		return false, err
+	}
+
+	return tableCount != 0, nil
+}
+
+// migration copies rows of the old key value table to the kine
+// table ONLY IF the key value table exists and migration has not been
+// done already. It creates the kine table if it doesn't exist.
+func migration(ctx context.Context, txn *sql.Tx, d *generic.Generic) error {
+	userVersionSQL := `PRAGMA user_version`
+	row := txn.QueryRowContext(ctx, userVersionSQL)
+
+	var userVersion int
+	if err := row.Scan(&userVersion); err != nil {
+		return err
+	}
+	// No need for migration - marker has already been set
+	if userVersion == 1 {
+		return nil
+	}
+
+	if userVersion > 1 {
```

Comment:
Lets suppose we ship 1.27 with user_version 1 and 1.28 with user_version 2. Lets also suppose we have a cluster of 5 nodes in 1.27 (user_version 1), three of them would be part of the quorum and two standby. If we upgrade a single node to 1.28 the user_version will switch to 2 and all remaining nodes in the cluster would be failing the call NewVariant -> setup -> migration. So at this point there is no quorum and if for any reason the cluster stops it will not come back up. This does not make me very comfortable. Can we avoid this behavior?

Reply:
I suppose this means that 1.28 _MUST NOT_ default to "user_version 2" without user intervention.

Reply:
If the schema version is newer, I think it makes sense that this returns an error, we are running a kine with an outdated schema and should fail here.

Reply:
@neoaggelos if the schema version is newer we must do something, indeed. Risking to get into an unrecoverable situation with the cluster down is not a good option. There are a number of ways to approach this problem. Lets discuss this before we assume/implement a strategy.